### DISCUSSION TITLE
About how biigle's new module interacts with VueJS

### DISCUSSION THREAD

**Discussion by Zuoyu2524** [biigle/core#617] [Q&A]
I tried many examples of common Blade templates interacting with Vue JS, but was not able to successfully bring up the desired interface, and the approach I used was as follows:
<div id="app">
            
        </div>
        <script src="{{ asset('js/app.js') }}"></script>
I found that maybe it is another way to use Vue in Biigle module when I read the codes  in MAIA but I couldn't understand  clearly the codes, Can you give me some instructions and help? Thank you!

**Comment by mzur** [comment]
Can you please describe again what interface you would like to have in the end? I vaguely remember that you want to add a new button to the volume overview and that there was a simple solution using the image label features of the volume overview and a more complex solution with entirely new views. Once I know again what you want, I'll describe how you can implement it.

Did you also have a look at the [tutorial](https://biigle-admin-documentation.readthedocs.io/module-development/module-development/) about module develoment?

**Comment by Zuoyu2524** [comment]
Ok, no problem, The page layout I want is similar to the Biigle annotation interface, the left is the picture display area, the right is the label selection area, after the label selection will appear a picture backward sliding button, the page I wrote a little with Vue roughly, almost like this: https://4frcy2-5173.csb.app/

**Reply by mzur** [reply]
I see. If I remember correctly then you planned to integrate a deep learning method for substrate classification into BIIGLE. It could work as simply as pressing a button (in the volume overview), then the DL model does its magic and returns classification results for each image as image labels. Is this correct?

Your description now sounds more like a new kind of annotation tool. Is this right? Your description and the example sounds very similar to what's currently possible with image labels in BIIGLE (both in the volume overview and the image annotation tool).

**Comment by Zuoyu2524** [comment]
Of course, this page in URI is just an example to explain the elements in the page, and it will use the interface style that incorporates Biigle later

**Comment by Zuoyu2524** [comment]
Ah yes, for the completeness of the module functionality, we added a training function that allows the user to manually label the images to train the network again and improve the generalization ability of the network. This is called label selection because our CNN can only recognize the labels specified by the network. I put all these specified labels into a label tree, and when the user chooses to use the training function for the labeling session, the system will automatically retrieve the contents of this label tree to show to the user for them to select the appropriate labels from these labels. for the image annotation, as you can see in the example I provided

**Reply by mzur** [reply]
Should the training be also done "inside" BIIGLE (like MAIA does)? In our meeting I got the impression that a retraining is not planned/possible.

Can't you use the image labelling tools of BIIGLE to generate new training data? This way you don't have to implement an entirely new tool.

I suggest to target a "minimum viable product" (MVP) for the module at first. This means that you can press a button in BIIGLE that starts the classification method. The classification results are attached as image labels to the images. When the classification task is finished, the user receives an email notification. This will already be lots of work. Once the MVP is ready, you can still continue and extend it with new features/views/tools.

**Comment by Zuoyu2524** [comment]
Yeah, I agree with you, I will start firstly with MVP. Thank you!

**Reply by mzur** [reply]
Sounds good! Please feel free to ask any questions about how to implement this. Most of what you need is already implemented in the [MAIA module](https://github.com/biigle/maia) so you will be able to copy lots of code from there.

**Comment by Zuoyu2524** [comment]
Hello,

I'm currently facing an issue with integrating Laravel and Python, specifically regarding how Laravel can determine if Python scripts have completed execution. Currently, my tasks are running continuously, and I am unable to proceed to the next step. However, I can see the results in a JSON file located in the "tmp" path.
Thank you for your help

**Reply by mzur** [reply]
Here is an example how Python is [called in MAIA](https://github.com/biigle/maia/blob/e6bb111590ff55c95bb4401aab6948730e353b00/src/Jobs/JobRequest.php#L168-L182):

```php
protected function python($command, $log = 'log.txt')
{
    $code = 0;
    $lines = [];
    $python = config('maia.python', '/usr/bin/python3');
    $logFile = "{$this->tmpDir}/{$log}";
    exec("{$python} -u {$command} >{$logFile} 2>&1", $lines, $code);

    if ($code !== 0) {
        $lines = File::get($logFile);
        throw new Exception("Error while executing python command '{$command}':\n{$lines}", $code);
    }

    return $logFile;
}
```

The method itself does not expect any input arguments or return any ouput from the Python script. Instead, the PHP code generates input JSON files, passes the paths to the files as command line arguments and the Python script generates output JSON files at locations specified in the input JSON files, so PHP can read them when the command is finished.

**Comment by Zuoyu2524** [comment]
The python script are very simple like this, it is just for a test.
with open(sys.argv[1]) as outfile:
    params = json.load(outfile)

images = params['images']
points = [1]
clusters = [images]
tmp_dir = params['tmp_dir']
for i, cluster in enumerate(clusters):
    for i, image in enumerate(cluster):
        image_path = '{}/{}.{}'.format(tmp_dir, image, 'json')
        with open(image_path, 'w') as outfile:
            json.dump(points, outfile)

**Comment by Zuoyu2524** [comment]
Ok, I understand, thank you

**Comment by Zuoyu2524** [comment]
Hello,
I am sorry, I want to know if python scripts has not errors, the log.txt is empty, is it normal?

**Comment by Zuoyu2524** [comment]
because the program is still at "Label recognition in progress.
Please come back later. "page and even I have already added the codes like maia to uploadthe results and change the jod state, but nothing helps

**Comment by Zuoyu2524** [comment]
protected function parseLabels($images)
    {
        $labels = [];
        $isNull = 0;

        // This might happen for corrupt image files which are skipped.

        foreach ($images as $image) {
            $path = "{$this->tmpDir}/{$image->getId()}.json";
            if (!File::exists($path)) {
                throw new Exception("Error while executing python command '{$command}':\n{$lines}", $code);
            }
            $newLabels = json_decode(File::get($path), true);
            if (is_null($newLabels)) {
                $isNull += 1;
            }
            $labels[] = $newLabels;
        }

        // For many images (as it is common) it might be not too bad if novelty detection
        // failed for some of them. We still get enough training proposals and don't want
        // to execute the long running novelty detection again. But if too many images
        // failed, abort.
        if (($isNull / count($images)) > 0.1) {
            throw new Exception('Unable to parse more than 10 % of the output JSON files.');
        }

        return $labels;
    }

**Comment by Zuoyu2524** [comment]
here is my codes to readthe file, I am not sure whether if it dosen't work, because these path for files already existed in 'tmp'

**Reply by mzur** [reply]
Can I see all the code in a repo somewhere? Also, please describe the steps you do (e.g. clicks in the UI, code executed manually) and what you expect as a result. While I understand that it might seem too much work for experiments and if you ae not that experienced with web development yet, I can only recommend to write tests for your (PHP) code. This way, you can test smaller parts of the codebase before you plug everything together. Errors are much easier to find in the smaller parts than in the whole thing.

Also, I'm on vacation for three weeks beginning tomorrow, so unfortunately I can't answer during that time.

**Comment by Zuoyu2524** [comment]
Hello,
I am sorry that I finished my job time yesterday and didn't see your message. Wish you a good holiday and tank you very much for your help, you can find my codes in github: https://github.com/Zuoyu2524/abysses. I will try to solve and wait for your reponse. Thank you again.

**Comment by Zuoyu2524** [comment]
Hello,

I'm happy to tell you that all the code for the interface and the database is working fine now, there is just one problem left, which is how to integrate the docker for the python environment used for label recognition into the application. Also, another problem is that the weight files for this network are up to 50 gigabytes, so maybe we need to find a place on the server to store them. In my localhost biigle application, the python executor is a very simple version that doesn't even run maia properly, and I'm just using very simple python code right now to make sure to run all the interfaces and databases without problems. So I would like to ask some questions about python environment and docker integration, thank you very much for your answer. In the meantime, you can find all the updated code at the github link: https://github.com/Zuoyu2524/abysses.

Wish you happy holidays!

**Reply by mzur** [reply]
MAIA has a storage path for weight files configured [like this](https://github.com/biigle/maia/blob/60ac67b28b47b9045035a57de3ab43199fbc70f5/src/config/maia.php#L90-L108). The files are loaded from the "urls" and stored locally at the "paths". In the job to perform the training, the files [are downloaded](https://github.com/biigle/maia/blob/60ac67b28b47b9045035a57de3ab43199fbc70f5/src/Jobs/ObjectDetectionRequest.php#L186-L196) if they are not already there.

If you want to execute the Python code locally, you can add the [GPU worker image](https://github.com/biigle/biigle/blob/8cc0752e0f86d90a303adb0bdae94aa9ee49cbee/build/gpu-worker.dockerfile) of the production configuration to your local `.docker` directory. Then add a section like this to `docker-compose.yml`:

```yaml
gpu-worker:
  image: biigle/gpu-worker
  build:
    context: ./.docker/
    dockerfile: gpu-worker.dockerfile
  user: ${USER_ID}:${GROUP_ID}
  depends_on:
    - app
    - database_testing
  tmpfs:
    - /tmp
  volumes_from:
    - app
  ipc: host # Required for Torch multiprocessing
  environment:
    - "DB_PORT=5432"
  init: true
  command: "php -d memory_limit=1G artisan queue:work --sleep=5 --tries=1 --timeout=0 --queue=gpu"
```

Run `docker compose build gpu-worker` to build the image and `docker compose up -d` afterwards to start the GPU worker. Finally, you have to make sure that your jobs are pushed to the `gpu` queue so they are picked up by the GPU worker and not the regular worker.

Please note that the Docker Compose configuration above does not actually expose the GPU to the Docker container (because I don't have one locally). So processing will be done on the CPU. If you want to use a local GPU you have to install NVIDIA Docker and add something [like this](https://github.com/biigle/biigle/blob/8cc0752e0f86d90a303adb0bdae94aa9ee49cbee/docker-compose.yaml#L55-L61) to your `docker-compose.yml`.

**Comment by Zuoyu2524** [comment]
Hello, ok, thank you very much, I will try this!

**Comment by Zuoyu2524** [comment]
Hello
I am sorry that I have try to use the gpu-worker, here will be the errors like this, and I change the version of php and also the dockerfile, but always the same errors
127.3 E: Unable to locate package php8.1-cli
127.3 E: Couldn't find any package by glob 'php8.1-cli'
127.3 E: Couldn't find any package by regex 'php8.1-cli'
127.3 E: Unable to locate package php8.1-curl
127.3 E: Couldn't find any package by glob 'php8.1-curl'
127.3 E: Couldn't find any package by regex 'php8.1-curl'
127.3 E: Unable to locate package php8.1-xml
127.3 E: Couldn't find any package by glob 'php8.1-xml'
127.3 E: Couldn't find any package by regex 'php8.1-xml'
127.3 E: Unable to locate package php8.1-pgsql
127.3 E: Couldn't find any package by glob 'php8.1-pgsql'
127.3 E: Couldn't find any package by regex 'php8.1-pgsql'
127.3 E: Unable to locate package php8.1-mbstring
127.3 E: Couldn't find any package by glob 'php8.1-mbstring'
127.3 E: Couldn't find any package by regex 'php8.1-mbstring'
127.3 E: Unable to locate package php8.1-redis
127.3 E: Couldn't find any package by glob 'php8.1-redis'
127.3 E: Couldn't find any package by regex 'php8.1-redis'

**Reply by mzur** [reply]
I got the same error just yesterday. This change will fix it: https://github.com/biigle/biigle/commit/89b92ad4fb14010e1c74fb65e30b9b063ed72a9b

**Comment by Zuoyu2524** [comment]
Hello,
Thank you! it's ok and may I ask you for some advice on how to blend our docker into biigle? Since our code is written by tensorflow 1.15.0, we may need to use a different docker than Maia, so should we write a new dockerfile after that to add the directive mapping to docker-compose.yml? Our docker link is as follows: https://hub.docker.com/r/psoto87/tf1.15.5-gpu

**Reply by mzur** [reply]
That's unfortunate, as BIIGLE recently moved away from TensorFlow. Now we use PyTorch as you probably discovered. Didn't we talk about this in our video call? I think so but I don't remember if we identified this problem then. TensorFlow 1 also has lots of security vulnerabilities and is quite outdated.

If you can't move away from TensorFlow you have to build your own Docker image. You also have to deploy it on a separate GPU (if you want to use the other GPU-based modules of BIIGLE, too, in your production instance).

[Here](https://github.com/biigle/biigle/blob/316fe0569b5b83af5cdfe01c7fc43af4a86f4ac3/build/gpu-worker.dockerfile) you can find the last MAIA Docker image that was based on TensorFlow 1 (from almost 3 years ago). Maybe you can use it as a starting point. But be warned that the TensorFlow and PHP versions as well as Python dependencies are outdated and probably have many security vulnerabilities!

**Comment by Zuoyu2524** [comment]
Hello,
Ok, I understand, we will communicate to upgrade the codes, Thank you!

**Comment by Zuoyu2524** [comment]
Hello, 

I have a little question that I can now successfully build the gpu-docker, but the peoblem is in biigle application, when running the scripts, the core didn't use the gpu-worker, so not deep learning environments, I try to make some changes to config but no success, the logs of docker show bellow:
```
(abysses) zhang@zhang-OptiPlex-7050:~/Biigle/biigle$ sudo docker-compose ps
[sudo] password for zhang: 
          Name                         Command               State                     Ports                   
---------------------------------------------------------------------------------------------------------------
biigle-app-1                docker-php-entrypoint php-fpm    Up      9000/tcp                                  
biigle-database-1           docker-entrypoint.sh postgres    Up      0.0.0.0:54320->5432/tcp,:::54320->5432/tcp
biigle-database_testing-1   docker-entrypoint.sh postgres    Up      5432/tcp                                  
biigle-gpu-worker-1         /opt/nvidia/nvidia_entrypo ...   Up      6006/tcp, 6064/tcp, 8888/tcp              
biigle-web-1                /docker-entrypoint.sh ngin ...   Up      0.0.0.0:8000->80/tcp,:::8000->80/tcp      
biigle-websockets-1         node /app/bin/server.js start    Up      6001/tcp                                  
biigle-worker-1             docker-php-entrypoint php  ...   Up                                                
(abysses) zhang@zhang-OptiPlex-7050:~/Biigle/biigle$ sudo docker-compose logs biigle-gpu-worker-1
ERROR: No such service: biigle-gpu-worker-1
(abysses) zhang@zhang-OptiPlex-7050:~/Biigle/biigle$ sudo docker-compose logs gpu-worker
Attaching to biigle-gpu-worker-1
biigle-gpu-worker-1 | 
biigle-gpu-worker-1 | ================
biigle-gpu-worker-1 | == TensorFlow ==
biigle-gpu-worker-1 | ================
biigle-gpu-worker-1 | 
biigle-gpu-worker-1 | NVIDIA Release 22.03-tf2 (build 33659467)
biigle-gpu-worker-1 | TensorFlow Version 2.8.0
biigle-gpu-worker-1 | 
biigle-gpu-worker-1 | Container image Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
biigle-gpu-worker-1 | Copyright 2017-2022 The TensorFlow Authors.  All rights reserved.
biigle-gpu-worker-1 | 
biigle-gpu-worker-1 | Various files include modifications (c) NVIDIA CORPORATION & AFFILIATES.  All rights reserved.
biigle-gpu-worker-1 | 
biigle-gpu-worker-1 | This container image and its contents are governed by the NVIDIA Deep Learning Container License.
biigle-gpu-worker-1 | By pulling and using the container, you accept the terms and conditions of this license:
biigle-gpu-worker-1 | https://developer.nvidia.com/ngc/nvidia-deep-learning-container-license
biigle-gpu-worker-1 | 
biigle-gpu-worker-1 | WARNING: The NVIDIA Driver was not detected.  GPU functionality will not be available.
biigle-gpu-worker-1 |    Use the NVIDIA Container Toolkit to start this container with GPU support; see
biigle-gpu-worker-1 |    https://docs.nvidia.com/datacenter/cloud-native/ .
biigle-gpu-worker-1 | 
(abysses) zhang@zhang-OptiPlex-7050:~/Biigle/biigle$ sudo docker-compose logs worker
Attaching to biigle-worker-1
biigle-worker-1     |   2023-08-16 12:27:21 Biigle\Modules\abysses\Listeners\DispatchAbyssesJob  RUNNING
biigle-worker-1     |   2023-08-16 12:27:21 Biigle\Modules\abysses\Listeners\DispatchAbyssesJob  277.88ms DONE
biigle-worker-1     |   2023-08-16 12:27:21 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  RUNNING
biigle-worker-1     |   2023-08-16 12:27:21 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  218.29ms FAIL
biigle-worker-1     |   2023-08-16 12:27:22 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  RUNNING
biigle-worker-1     |   2023-08-16 12:27:22 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  94.46ms FAIL
biigle-worker-1     |   2023-08-16 12:27:22 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  RUNNING
biigle-worker-1     |   2023-08-16 12:27:22 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  77.44ms FAIL
biigle-worker-1     |   2023-08-16 12:27:22 Biigle\Modules\abysses\Jobs\LabelRecognitionFailure  RUNNING
biigle-worker-1     |   2023-08-16 12:27:24 Biigle\Modules\abysses\Jobs\LabelRecognitionFailure  2,315.04ms DONE
biigle-worker-1     |   2023-08-16 12:28:54 Biigle\Modules\abysses\Listeners\DispatchAbyssesJob  RUNNING
biigle-worker-1     |   2023-08-16 12:28:54 Biigle\Modules\abysses\Listeners\DispatchAbyssesJob  47.95ms DONE
biigle-worker-1     |   2023-08-16 12:28:55 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  RUNNING
biigle-worker-1     |   2023-08-16 12:28:55 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  232.28ms FAIL
biigle-worker-1     |   2023-08-16 12:28:55 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  RUNNING
biigle-worker-1     |   2023-08-16 12:28:56 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  658.33ms FAIL
biigle-worker-1     |   2023-08-16 12:28:56 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  RUNNING
biigle-worker-1     |   2023-08-16 12:28:57 Biigle\Modules\abysses\Jobs\LabelRecognitionRequest  743.35ms FAIL
biigle-worker-1     |   2023-08-16 12:28:57 Biigle\Modules\abysses\Jobs\LabelRecognitionFailure  RUNNING
biigle-worker-1     |   2023-08-16 12:28:58 Biigle\Modules\abysses\Jobs\LabelRecognitionFailure  390.66ms DONE
```

**Reply by mzur** [reply]
It says that it could not detect a GPU. Is this correct? Otherwise you may have to install the NVIDIA Container Toolkit. Also make sure that you have a resource definition like [this one](https://github.com/biigle/biigle/blob/230c5be6f4a9bc6ae0cca036def1a96974c60619/docker-compose.yaml#L55-L61) for the service.

Usually TensorFlow falls back to CPU computation without a GPU so this may or may not be the cause of the error. A failed job probably puts error messages in the `storage/logs/` directory. Please have a look there (and maybe upload the log file(s) here).

**Reply by mzur** [reply]
I now saw that the `LabelRecognitionRequest` jobs were executed by the `worker` service instead of the `gpu-worker` service. Make sure that the jobs are submitted to the `gpu` queue and that the service process listens to this queue with `--queue gpu` ([example](https://github.com/biigle/biigle/blob/230c5be6f4a9bc6ae0cca036def1a96974c60619/docker-compose.yaml#L50) with different queue names).

In some setups I used a `gpu` *connection* instead or in addition to a `gpu` queue. This looks like [this](https://github.com/biigle/biigle/blob/230c5be6f4a9bc6ae0cca036def1a96974c60619/docker-compose.yaml#L68) where the connection name is the last argument of the `queue:work` command. This is *not* what you need here unless you explicitly configure a new connection in the queue config file.

**Comment by Zuoyu2524** [comment]
Yes, you're right, I really don't have a GPU configuration on my computer at the moment, that should be the problem, thank you!

**Reply by mzur** [reply]
I don't know, TensorFlow usually falls back to the CPU in that case. Please read my second comment [regarding queues](https://github.com/orgs/biigle/discussions/617#discussioncomment-6741582).
